File: c/problem_23.c
#include "problem_23.h"
#include <stdio.h>
#include <string.h>
#include <assert.h>

#define MAX_CELLS 256

typedef struct { int r, c, steps; } Cell;

/* shortestPath: BFS. O(M*N) time and space. */
int shortestPath(int rows, int cols, int grid[rows][cols],
                 int startR, int startC, int endR, int endC) {
    if (grid[startR][startC] || grid[endR][endC]) return -1;

    int visited[16][16];
    memset(visited, 0, sizeof(visited));
    visited[startR][startC] = 1;

    Cell queue[MAX_CELLS];
    int front = 0, back = 0;
    queue[back++] = (Cell){startR, startC, 0};

    int dr[] = {-1,1,0,0}, dc[] = {0,0,-1,1};
    while (front < back) {
        Cell cur = queue[front++];
        if (cur.r == endR && cur.c == endC) return cur.steps;
        for (int d = 0; d < 4; d++) {
            int nr = cur.r + dr[d], nc = cur.c + dc[d];
            if (nr >= 0 && nr < rows && nc >= 0 && nc < cols
                && !grid[nr][nc] && !visited[nr][nc]) {
                visited[nr][nc] = 1;
                queue[back++] = (Cell){nr, nc, cur.steps + 1};
            }
        }
    }
    return -1;
}

void problem23(void) {
    printf("Problem 23:\n");
    int grid[4][4] = {
        {0,0,0,0},
        {1,1,0,1},
        {0,0,0,0},
        {0,0,0,0},
    };
    assert(shortestPath(4, 4, grid, 3, 0, 0, 0) == 7);
    assert(shortestPath(4, 4, grid, 0, 0, 0, 0) == 0);
    int blocked[2][2] = {{0,1},{1,0}};
    assert(shortestPath(2, 2, blocked, 0, 0, 1, 1) == -1);
    printf("All Tests Passed\n");
}

File: rust/problem_23.rs
use std::collections::VecDeque;

// shortest_path returns the minimum steps from start to end in a boolean grid
// where true = wall. Uses BFS. Returns None if no path exists.
// This runs in O(M*N) time and space.
pub fn shortest_path(
    grid: &[Vec<bool>],
    start: (usize, usize),
    end: (usize, usize),
) -> Option<usize> {
    let rows = grid.len();
    let cols = grid[0].len();
    if grid[start.0][start.1] || grid[end.0][end.1] { return None; }

    let mut visited = vec![vec![false; cols]; rows];
    visited[start.0][start.1] = true;
    let mut queue: VecDeque<(usize, usize, usize)> = VecDeque::new();
    queue.push_back((start.0, start.1, 0));

    while let Some((r, c, steps)) = queue.pop_front() {
        if (r, c) == end { return Some(steps); }
        for (dr, dc) in [(-1i32,0),(1,0),(0,-1i32),(0,1)] {
            let nr = r as i32 + dr;
            let nc = c as i32 + dc;
            if nr >= 0 && nr < rows as i32 && nc >= 0 && nc < cols as i32 {
                let (nr, nc) = (nr as usize, nc as usize);
                if !grid[nr][nc] && !visited[nr][nc] {
                    visited[nr][nc] = true;
                    queue.push_back((nr, nc, steps + 1));
                }
            }
        }
    }
    None
}

pub fn problem23() {
    println!("Problem 23:");
    let f = false; let t = true;
    let grid = vec![
        vec![f,f,f,f],
        vec![t,t,f,t],
        vec![f,f,f,f],
        vec![f,f,f,f],
    ];
    assert_eq!(shortest_path(&grid, (3,0), (0,0)), Some(7), "assertion failed");
    assert_eq!(shortest_path(&grid, (0,0), (0,0)), Some(0), "assertion failed");
    let blocked = vec![vec![f,t],vec![t,f]];
    assert_eq!(shortest_path(&blocked, (0,0), (1,1)), None, "assertion failed");
    println!("All Tests Passed");
}
